LLVM JIT interface

pyos opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

This pull request adds the capability to compile built-in functions with numeric (possibly nullable) arguments and return type to native code through `llvm::IRBuilder`. Compilable functions are automatically inlined into each other for a performance boost. The function can then also decide where to evaluate each of its inlined arguments, allowing for some laziness (non-compilable subexpressions still have to be evaluated eagerly).

This PR also includes implementations of the interface for most arithmetic and logic functions. Don't have any performance comparisons yet, though.

Known problems:

 * ~~this code will break with LLVM 7 when it's released due to API incompatibilities;~~
 * LLVM has a ton of unused parameters in its header files, so I had to add `-Wno-unused-parameter` to clickhouse_functions;
 * *always* inlining compilable subexpressions completely undoes all of common subexpression eliminator's work — using a heuristic, e.g. thresholding based on the size of the subexpression's graph multiplied by the number of times it is used, might be better overall;
 * `and` has somewhat weird semantics in that `and(false, null)` is null. This means it's impossible for `and` to be lazy in the second argument if it's nullable. (`and(x, non-nullable)` and `and(null, x)` work fine, though.)

pyos committed:
Add an example function that uses LLVM to compile its own body

pyos committed:
Add a JIT interface for row-wise default-nullable functions.

Not actually implemented, though. It does print out some jit-compiled stuff,
but that's about it. For example, this query:

    select number from system.numbers where something(cast(number as Float64)) == 4

results in this on server's stderr:

    define double @"something(CAST(number, 'Float64'))"(void**, i8*, void*) {
    "something(CAST(number, 'Float64'))":
      ret double 1.234500e+04
    }

(and an exception, because that's what the non-jitted method does.)

As one may notice, this function neither reads the input (first argument;
tuple of arrays) nor writes the output (third argument; array), instead
returning some general nonsense.

In addition, `#if USE_EMBEDDED_COMPILER` doesn't work for some reason,
including LLVM headers requires -Wno-unused-parameter, this probably only
works on LLVM 5.0 due to rampant API instability, and I'm definitely
no expert on CMake. In short, there's still a long way to go.

pyos committed:
Map all number types to LLVM types.

The example from the previous commit doesn't need a cast to Float64 anymore.

pyos committed:
Implement JIT compilation, without a loop for now.

It actually seems to work, so long as you only have one row that is. E.g.

    > select something(cast(number + 6 as Float64), cast(number + 2 as Float64)) from system.numbers limit 1';
    8

with this IR:

    define void @"something(CAST(plus(number, 6), 'Float64'), CAST(plus(number, 2), 'Float64'))"(void**, i8*, double*) {
    entry:
      %3 = load void*, void** %0
      %4 = bitcast void* %3 to double*
      %5 = load double, double* %4
      %6 = getelementptr void*, void** %0, i32 1
      %7 = load void*, void** %6
      %8 = bitcast void* %7 to double*
      %9 = load double, double* %8
      %10 = fadd double %5, %9
      store double %10, double* %2
      ret void
    }

pyos committed:
Separate jit-compilability checks from actual compilation

pyos committed:
Implement a loop over the columns in jit-compiled code

pyos committed:
Extend the test jit-compilable function to arbitrary numbers

pyos committed:
Enable the default set of LLVM optimization passes

I honestly can't tell if they work. LLVM has surprisingly bad API documentation.

pyos committed:
Remove IFunction::createResultColumn.

Given that the list of supported types is hardcoded in
LLVMContext::Data::toNativeType, this method is redundant because
LLVMPreparedFunction can create a ColumnVector itself.

pyos committed:
Inline jit-compilable functions into other jit-compilable functions

pyos committed:
Tweak the jit compilation API to be more amenable to lazy computation

pyos committed:
Add missing option to CMakeFiles.txt

pyos committed:
Fix CMakeFiles syntax

pyos committed:
Fix some comments

pyos committed:
Support nullable columns (with default behavior) in jitted functions

pyos committed:
Do not compile the jit if USE_EMBEDDED_COMPILER is disabled

pyos committed:
Fix some comments' style

pyos committed:
Move function compilation before insertion of REMOVE_COLUMNs

pyos committed:
Implement informational methods for LLVMFunction

pyos committed:
Poison only columns actually used by ARRAY_JOIN and JOIN

KochetovNicolai commented on dbms/src/Interpreters/ExpressionActions.cpp:
```diff
@@ -990,6 +994,96 @@ void ExpressionActions::optimizeArrayJoin()
+    std::vector<bool> redundant(actions.size());
+    // an empty optional is a poisoned value prohibiting the column's producer from being removed
+    // (which it could be, if it was inlined into every dependent function).
+    std::unordered_map<std::string, std::unordered_set<std::optional<size_t>>> current_dependents;
```
It's better use struct `{ bool is_used; std::unordered_set<size_t> dependents; }` instead of `std::unordered_set<std::optional<size_t>>`

pyos committed:
Remove an unnoticed debug `return`

proller commented on dbms/src/Server/Compiler-5.0.0/CMakeLists.txt:
```diff
@@ -51,3 +51,8 @@ libtinfo.a
+
+if (MAKE_STATIC_LIBRARIES)
+    # fix strange static error: undefined reference to 'std::error_category::~error_category()'
+    target_link_libraries(clickhouse-compiler-lib PUBLIC stdc++)
```
This fix nedded only for old distributions (ubuntu trusty, xenial)
In ubuntu artful anything ok.
ant this fix adds dependency to shared system lib - it unacceptable for our fully static package.

proller commented on dbms/CMakeLists.txt:
```diff
@@ -82,6 +82,15 @@ list (APPEND dbms_headers
+    # TODO: global-disable no-unused-parameter
+    set_source_files_properties(src/Interpreters/ExpressionJIT.cpp PROPERTIES COMPILE_FLAGS "-Wno-unused-parameter -Wno-non-virtual-dtor")
+else ()
+    list (REMOVE_ITEM dbms_sources src/Interpreters/ExpressionJIT.cpp)
```
use in this files:
```
#include <Common/config.h>
#if USE_EMBEDDED_COMPILER
...
#endif
```

pyos commented on dbms/src/Interpreters/ExpressionJIT.cpp:
```diff
@@ -0,0 +1,318 @@
+            /// assume the column is a `ColumnVector<T>`. there's probably no good way to actually
+            /// check that at runtime, so let's just hope it's always true for columns containing types
+            /// for which `LLVMContext::Data::toNativeType` returns non-null.
+            columns[i] = column->getDataAt(0).data;
```
`IColumn::isFixedAndContiguous` seems to be what I wanted here. The compiled loop could be extended to arbitrary columns for which this method returns true passing a tuple of strides (e.g. string length for `ColumnFixedString`) instead of a tuple of "is constant" flags, though I'm not sure how this could affect loop auto-vectorization.

pyos committed:
Add IColumn::getRawData to fixed-contiguous columns

pyos committed:
Put #if USE_EMBEDDED_COMPILER in ExpressionJIT.{cpp,h}

pyos committed:
Minor style fixes

pyos committed:
Throw in untyped versions of IFunction::{isCompilable,compile}

IFunction inherits IFunctionBase for some reason despite not actually knowing
the types, so these two methods make no sense. The versions with DataTypes&
as an argument should be used instead.

pyos committed:
Inline compile-time constants into jitted functions.

pyos committed:
Use system LLVMConfig.cmake with minor tweaks.

Should fix Travis build, finally.

pyos committed:
Refactor LLVMFunction to make extending to DataTypeNullable easier

pyos committed:
Let jit-compilable functions deal with NULLs themselves.

And provide a default implementation of compile() for nullable columns
that actually works and is consistent with execute().

pyos committed:
Check nativity of all types *before* calling isCompilable

pyos committed:
Represent nullable types as pairs instead of pointers.

Turns out LLVM has insertvalue & extractvalue for struct in registers. This is
faster than pointers because null checks are now subject to more optimizations.

pyos committed:
Inline nullable number constants into compiled code.

Also, protect against some segfaults during compilation by checking
correctness of the type returned by compile().

pyos committed:
Make LLVMFunction monotonicity computation shorter (and fix a typo-bug)

pyos committed:
compilePrologue() isn't particularly useful after all.

Basically the only thing it can do that compile() can't is create 'alloca'
instructions, which are only needed to get pointers to stack variables.
Given that dynamically-sized allocations aren't possible with this API,
such pointers are probably completely pointless (heh).

pyos committed:
Support {Date,DateTime,Interval,UUID,FixedString} in compiled functions

pyos committed:
Move ExpressionActions::compileFunctions to ExpressionJIT.cpp.

This means ExpressionJIT.h only has to expose one function.

pyos committed:
Hide the whole JIT API behind #if USE_EMBEDDED_COMPILER

Kind ugly, but at least the conditionals are used consistently now.

pyos committed:
Remove getDefaultNativeValue in favor of llvm::Constant::getNullValue

pyos committed:
Extract the code that compiles a single IFunctionBase from LLVMFunction

pyos committed:
Merge branch 'master' of https://github.com/yandex/ClickHouse into llvm-jit

pyos committed:
Fix a bug that limited inlining depth at 2

pyos committed:
Implement jit for most arithmetic functions, remove the test function

pyos committed:
Work around a bug in llvm::IRBuilder::CreateMaxNum

pyos committed:
Remove outdated comments

pyos committed:
Remove laziness on nullable arguments from default implementation

It breaks semantics, sadly.

pyos committed:
Implement jit for logic functions

pyos committed:
Merge branch 'master' of https://github.com/yandex/ClickHouse into llvm-jit

pyos committed:
Merge API changes from upstream

pyos committed:
Add a setting that disables jit-compilation

ztlpn commented:
autotests

pyos committed:
Remove dynamically-linked stdc++ from dependencies

pyos committed:
If all inputs to a jitted function are constant, return a constant

pyos committed:
Don't waste time jit-compiling isolated functions.

This is already done ahead of time when building the executable.

ztlpn commented:
autotests

alexey-milovidov commented:
> always inlining compilable subexpressions completely undoes all of common subexpression eliminator's work — using a heuristic, e.g. thresholding based on the size of the subexpression's graph multiplied by the number of times it is used, might be better overall;

That's Ok. The effect is like to be neglible because common subexpressions of compileable expressions are expected to be cheap or if a whole compileable expression contains multiple identical subexpressions, it will be optimized by compiler.

> LLVM has a ton of unused parameters in its header files, so I had to add -Wno-unused-parameter to clickhouse_functions;

You can enable it only around #include with #pragma GCC diagnostic.

> and has somewhat weird semantics in that and(false, null) is null. This means it's impossible for and to be lazy in the second argument if it's nullable. (and(x, non-nullable) and and(null, x) work fine, though.)

Ok.

Commit:
Merge branch 'llvm-jit' of https://github.com/pyos/ClickHouse into pyos-llvm-jit

pyos committed:
Resolve symbols right after compiling.

llvm::orc::RTDyldObjectLinkingLayer::findSymbol appears to be non-threadsafe.

alexey-milovidov committed:
Merge branch 'llvm-jit' of https://github.com/pyos/ClickHouse into pyos-llvm-jit

alexey-milovidov committed:
Make warning suppressions more local #2277

alexey-milovidov committed:
Removed deprecated CMake function #2277

alexey-milovidov committed:
Miscellaneous #2277

alexey-milovidov committed:
Addition to prev. revision #2277

alexey-milovidov commented:
TODO:
- implement for bitwise, comparison, conditional functions with numeric argument if not already;
- compare performance of integer division by constant with current implementation (that is using libdivide);
- compare performance of embedded simple hash functions (intHash32, intHash64) with implementation in SQL;
- compare performance for random forest embedded in SQL, like this: https://gist.githubusercontent.com/alexey-milovidov/000d340779f9ce0653df97a85887a3a0/raw/d70cac3fa2868462d32ac043c7eb843105badd12/forest_in_sql.sql
- measure overhead of compilation;
- investigate the need for caching of compiled functions and measure memory usage;
- implement a setting to disable compilation;

alexey-milovidov commented on dbms/src/Interpreters/ExpressionJIT.cpp:
```diff
@@ -0,0 +1,525 @@
+        builder.populateFunctionPassManager(fpm);
+        for (auto & function : *module)
+            fpm.run(function);
+        llvm::cantFail(compileLayer.addModule(module, std::make_shared<llvm::orc::NullResolver>()));
```
Better to throw exception.

alexey-milovidov committed:
Miscellaneous #2277

alexey-milovidov committed:
Allow to build with clang 7

alexey-milovidov committed:
Fixed code #2272

alexey-milovidov committed:
Fixed code #2272

alexey-milovidov committed:
Fixed code #2272

alexey-milovidov committed:
Allow to link with LLVM if it was compiled without RTTI #2277

alexey-milovidov commented on cmake/find_llvm.cmake:
```diff
@@ -9,7 +9,7 @@ if (ENABLE_EMBEDDED_COMPILER)
     if (LLVM_FOUND)
         # Remove dynamically-linked zlib and libedit from LLVM's dependencies:
-        set_target_properties(LLVMSupport PROPERTIES INTERFACE_LINK_LIBRARIES "-lpthread;LLVMDemangle;stdc++")
+        set_target_properties(LLVMSupport PROPERTIES INTERFACE_LINK_LIBRARIES "-lpthread;LLVMDemangle")
```
It doesn't removed for whatever reason, even more: we have many new dependencies (all of them must be linked statically).

```
$ ldd ./clickhouse
        linux-vdso.so.1 =>  (0x00007ffd55383000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f1a7ce24000)
        libLTO.so.7 => /usr/local/lib/libLTO.so.7 (0x00007f1a79e1f000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f1a79c17000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f1a79a13000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f1a7970a000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f1a79341000)
        /lib64/ld-linux-x86-64.so.2 (0x000055ff1da45000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f1a79127000)
        libtinfo.so.5 => /lib/x86_64-linux-gnu/libtinfo.so.5 (0x00007f1a78efe000)
        libc++.so.1 => /usr/local/lib/../lib/libc++.so.1 (0x00007f1a78c33000)
        libc++abi.so.1 => /usr/local/lib/../lib/libc++abi.so.1 (0x00007f1a789fd000)
        libgcc_s.so.1 => /usr/local/lib64/libgcc_s.so.1 (0x00007f1a787e5000)
```

pyos replied:
Can't reproduce this, which version of LLVM are you using? Perhaps something is wrong with Compiler-X.Y.Z/CMakeLists.txt?

alexey-milovidov commented on dbms/src/Functions/FunctionsArithmetic.h:
```diff
@@ -264,9 +348,19 @@ struct BitShiftLeftImpl
+    static inline llvm::Value * compile(llvm::IRBuilder<> & b, llvm::Value * left, llvm::Value * right, bool)
+    {
+        if (!left->getType()->isIntegerTy())
+            throw Exception("BitShiftLeftImpl expected an integral type", ErrorCodes::LOGICAL_ERROR);
```
It doesn't. Example:

```
SELECT bitShiftLeft(1.1, number) FROM system.numbers LIMIT 10
```

pyos replied:
That simply casts 1.1 to an integer type, which is done by `FunctionBinaryArithmetic::compileImpl`. *This* method should never receive a non-int type.

alexey-milovidov commented on dbms/src/Functions/FunctionsArithmetic.h:
```diff
@@ -293,6 +397,19 @@ struct BitRotateLeftImpl
+        if (!left->getType()->isIntegerTy())
+            throw Exception("BitRotateLeftImpl expected an integral type", ErrorCodes::LOGICAL_ERROR);
+        auto * size = llvm::ConstantInt::get(left->getType(), left->getType()->getPrimitiveSizeInBits());
+        /// XXX how is this supposed to behave in signed mode?
```
Any reasonable implementation specific behaviour.
PS. Doesn't LLVM contain "rotate" instructions?

pyos replied:
The problem here is that only `execute` is implementation-defined, while behavior of `compile` depends only on whether `lshr` or `ashr` is used. This means enabling compilation may change behavior.

>PS. Doesn't LLVM contain "rotate" instructions?

Not any I know of. The only reference I could find is a discussion about MIPS `rotr` instructions on the llvm-dev mailing list, where the bottom line is that a dedicated "rotate" instruction was never added to LLVM because it would increase the optimizer's complexity.

alexey-milovidov commented on dbms/src/Functions/FunctionsArithmetic.h:
```diff
@@ -463,6 +656,10 @@ struct IntExp2Impl
     }
+
+#if USE_EMBEDDED_COMPILER
+    static constexpr bool compilable = false; /// library function
```
This is trivial, please look inside.

pyos replied:
Yeah, should've guessed...

alexey-milovidov committed:
Fixed style a bit [#CLICKHOUSE-2]

alexey-milovidov committed:
Applied clang-format to some imported code [#CLICKHOUSE-2]

alexey-milovidov committed:
Applied clang-format #2272

alexey-milovidov committed:
Fixed style [#CLICKHOUSE-2]

alexey-milovidov committed:
Applied clang-format to some imported code [#CLICKHOUSE-2]

alexey-milovidov committed:
Applied clang-format #2272

alexey-milovidov committed:
Style [#CLICKHOUSE-2]

alexey-milovidov committed:
Style [#CLICKHOUSE-2]

alexey-milovidov committed:
Added a tool to fix code style [#CLICKHOUSE-2]

alexey-milovidov committed:
Addition to prev. revision [#CLICKHOUSE-2]

alexey-milovidov commented:
compileFunctions is executed multiple times for the same expression. For example, for query:
```
SELECT count() FROM system.numbers_mt WHERE NOT ignore(number + number)
```
it is executed 8 times.

alexey-milovidov commented:
We need at least one functional test that proves that compilation actually works.

alexey-milovidov committed:
Fixed error; added ProfileEvent #2277

alexey-milovidov commented:
Queries for hash function:

```
WITH
    bitXor(number, 0x4CF2D2BAAE6DA887) AS x0,
    bitXor(x0, bitShiftRight(x0, 33)) AS x1,
    x1 * 0xff51afd7ed558ccd AS x2,
    bitXor(x2, bitShiftRight(x2, 33)) AS x3,
    x3 * 0xc4ceb9fe1a85ec53 AS x4,
    bitXor(x4, bitShiftRight(x4, 33)) AS x5
SELECT x5, intHash64(number) FROM system.numbers LIMIT 10;

SELECT count() FROM system.numbers WHERE NOT ignore(intHash64(number));

WITH
    bitXor(number, 0x4CF2D2BAAE6DA887) AS x0,
    bitXor(x0, bitShiftRight(x0, 33)) AS x1,
    x1 * 0xff51afd7ed558ccd AS x2,
    bitXor(x2, bitShiftRight(x2, 33)) AS x3,
    x3 * 0xc4ceb9fe1a85ec53 AS x4,
    bitXor(x4, bitShiftRight(x4, 33)) AS x5
SELECT count() FROM system.numbers WHERE NOT ignore(x5);
```

alexey-milovidov commented:
Compiled expressions are invisible in `perf top`. How we can manage to make them visible (at least under some single name)?

alexey-milovidov committed:
Better error checks #2277

alexey-milovidov committed:
Merge branch 'pyos-llvm-jit' of github.com:yandex/ClickHouse into pyos-llvm-jit

alexey-milovidov committed:
Build fixes #2277

alexey-milovidov committed:
Removed useless library [#CLICKHOUSE-2]

alexey-milovidov committed:
Merge branch 'pyos-llvm-jit' of github.com:yandex/ClickHouse into pyos-llvm-jit

alexey-milovidov committed:
Removed useless library [#CLICKHOUSE-2]

alexey-milovidov committed:
Merge remote-tracking branch 'origin/master' into pyos-llvm-jit

pyos commented:
>compileFunctions is executed multiple times for the same expression. For example, for query:

It is executed as many times as `ExpressionActions` is constructed. If that's a problem, perhaps `ExpressionActions` should be reused instead of walking the AST every time. This is probably outside the scope of this PR.

>We need at least one functional test that proves that compilation actually works.

The current tests should already do that with `USE_EMBEDDED_COMPILER=1` since the `compile_expressions` setting is enabled by default. This is true for the gcc-7 Travis run, for instance.

>Compiled expressions are invisible in perf top. How we can manage to make them visible (at least under some single name)?

Technically, they should be visible as `0x7f...`, although the address will always be different. I'm pretty sure the only way to give them a name is to pass a custom symbol map to `perf`.

alexey-milovidov commented:
> The current tests should already do that with USE_EMBEDDED_COMPILER=1 since the compile_expressions setting is enabled by default. This is true for the gcc-7 Travis run, for instance.

It's not obvious that compiled version is actually used.
I've added a ProfileEvent for this purpose. You can add a test with `SELECT ... FROM system.events` (or propose a better solution).

pyos committed:
Assume output and one input of LLVMFunction is non-const

pyos committed:
Fix addModule check for LLVM < 7

It returns `Expected`, so false-y is a failure. (In >= 7 it returns
an `Error`, so truth-y is a failure. Wow, that's confusing.)

pyos committed:
Remove a redundant target_compile_options

pyos committed:
Implement IntExp2Impl::compile

pyos commented:
>Queries for hash function:

Looks like pyos@0e09138 somehow makes the SQL version faster than the built-in function (400M rows/s vs. 360M). Huh.

alexey-milovidov committed:
Merge branch 'master' into pyos-llvm-jit

alexey-milovidov committed:
Merge branch 'master' of github.com:yandex/ClickHouse into pyos-llvm-jit

pyos committed:
Select an *exact* target machine, not an approximation.

Required for enabling advanced features such as AVX and AVX2.

Code mostly copied from LLVM's tools/opt/opt.cpp.

pyos committed:
Merge branch 'master' of https://github.com/yandex/ClickHouse into llvm-jit

Conflicts:
	debian/.pbuilderrc

pyos committed:
Copy changes from LLVM {5,6} CMakeLists to 7

alexey-milovidov committed:
Removed useless code [#CLICKHOUSE-2]

pyos committed:
Extract native bool cast; generalize number cast to nullables

pyos committed:
Implement jit for numeric if and multiIf

pyos commented:
>I've added a ProfileEvent for this purpose. You can add a test with SELECT ... FROM system.events (or propose a better solution).

Such a test would need to be skipped if `USE_EMBEDDED_COMPILER` is 0. Is that possible?

Commit:
Merge branch 'llvm-jit' of github.com:pyos/ClickHouse into pyos-llvm-jit

Commit:
Merge branch 'master' of github.com:yandex/ClickHouse into pyos-llvm-jit

alexey-milovidov commented:
> Such a test would need to be skipped if USE_EMBEDDED_COMPILER is 0. Is that possible?

Add "compile" word to the test name (we have two existing tests of that kind).
(As far as I remember, they aren't disabled automatically, but will be easily disabled.)

alexey-milovidov committed:
Merge branch 'master' of github.com:yandex/ClickHouse into pyos-llvm-jit

pyos committed:
Fix a typo

alexey-milovidov committed:
Merge branch 'master' of github.com:yandex/ClickHouse into pyos-llvm-jit

alexey-milovidov committed:
Fixed linking of tests #2277

alexey-milovidov committed:
Merge branch 'master' of github.com:yandex/ClickHouse into pyos-llvm-jit

alexey-milovidov committed:
Updated submodules #2277

alexey-milovidov committed:
Fixed error #2277

alexey-milovidov committed:
Removed hack #2277

alexey-milovidov committed:
Fixed include paths #2277

alexey-milovidov committed:
Merge branch 'pyos-llvm-jit' of github.com:yandex/ClickHouse into pyos-llvm-jit

alexey-milovidov committed:
Fixed include paths #2277

alexey-milovidov committed:
Updated submodule #2277

alexey-milovidov committed:
Updated submodule #2277

alexey-milovidov committed:
Updated submodule #2277

alexey-milovidov committed:
Updated submodule #2277

alexey-milovidov committed:
Fixed error #2277

alexey-milovidov committed:
Updated submodule #2277

alexey-milovidov committed:
Updated submodule #2277

pyos committed:
Merge branch 'pyos-llvm-jit' of https://github.com/yandex/ClickHouse into llvm-jit

pyos committed:
Merge branch 'master' of https://github.com/yandex/ClickHouse into llvm-jit

pyos committed:
Allow calling into libc from jitted code.

Mostly for intrinsics like memcpy/memset/memmove, which are inserted
during optimization by LLVM itself. (With a null resolver, a compiled
version of something like `Uint64 < 0` would segfault.)

pyos committed:
Implement jit for comparisons, except for (double, int).

That one has some edge cases which I can't be bothered to code.

pyos committed:
Fix incorrect phi node edges in `if`

pyos committed:
Bridge between incompatible LLVM APIs

pyos committed:
Revert the part of bd332b that moved a read after std::move.
